### PR TITLE
synthetic 'slot' and 'block' events to inform block boundaries

### PR DESCRIPTION
**Pull request by sunnygleason**
#### Problem

Current assumptions:

* users of block explorer think of transaction according to transaction_id and block_id
* block_id, block_height are not present when the entry+transaction is streamed
* entry_id, tick_height are a less important concept and should be less prominent in the ui
* we can store entries and transactions as they are streamed
* block_id is not known until the last entry in the block
* we write back the (entry -> block_id) and (block_id -> entry) relations after the block_id is streamed
* transactions get block_id by querying the block_id of their entry_id
* (we would otherwise have a high volume of transactions to individually update their block_id)

Open issue:

* what happens in the context of forks?

#### Summary of Changes

* when we write the last entry in the slot (block?), we additionally emit a "block" event
* all block parameters are stubbed - they may need to be different

### CODE REVIEW

**Comment by mvines** [comment]
Seems ok to me overall.   To manage forks I think we'd want to do something like the RPC API overall, emit an event when the "primary fork" (not a real definition yet) that the RPC API is following changes.   This all doesn't exist yet.

**Comment by CriesofCarrots** [comment]
@sunnygleason, this approach seems fine to me as well. Thanks for the detailed description of assumptions.
@mvines , to confirm, entries from all forks will reach the replay_stage, correct?

**Comment by mvines** [comment]
> to confirm, entries from all forks will reach the replay_stage, correct?
Not sure, I'm not following the WIP forking PRs that closely right now

**Comment by sunnygleason** [comment]
@mvines @CriesofCarrots is this ok to merge? And/or should I run it by
folks like @garious and @aeyakovenko or others who have
contemplated/are contemplating the forking model?

**Comment by mvines** [comment]
I’d like that const usage to be removed before merging, rest seems fine for now and we can iterate as needed

**Comment by mvines** [comment]
Bah, bank.get_current_leader returns the leader for the *next* tick/slot instead of *this* one.  Use it for now though - if this PR lands before mine, I’ll fix this.  If I win, you’ll get a merge conflict and I’ll help you through it

**Comment by sunnygleason** [comment]
> I'm thinking we may want to also include the current leader id in the block event.
> We can grab that from `self.bank` as well, and use this method:
> [solana/src/bank.rs](https://github.com/solana-labs/solana/blob/549f9676f1e144d17927121f84ef9382ea3239ad/src/bank.rs#L885)
> 
> Line 885 in [549f967](/solana-labs/solana/commit/549f9676f1e144d17927121f84ef9382ea3239ad)
> 
>  pub fn get_current_leader(&self) -> Option<(Pubkey, u64)> { 
> 
> I'd say go ahead and roll that into this PR.
> Also, @sunnygleason , can you edit the PR subject to remove RFC before merge?

@CriesofCarrots (1) I'll take a look & get that in, and (2) yes of course!

**Comment by mvines** [comment]
DEFAULT_TICKS_PER_SLOT removal:
```patch
From 1ed312ea143f99e2616f2c2a90c8c46faa8c62ce Mon Sep 17 00:00:00 2001
From: Michael Vines <mvines@gmail.com>
Date: Thu, 7 Feb 2019 13:51:34 -0800
Subject: [PATCH] Fetch ticks-per-slot from leader scheduler

---
 src/entry_stream.rs | 23 +++++++++++++++--------
 src/replay_stage.rs |  4 +++-
 2 files changed, 18 insertions(+), 9 deletions(-)

diff --git a/src/entry_stream.rs b/src/entry_stream.rs
index 19287788..4fbe4438 100644
--- a/src/entry_stream.rs
+++ b/src/entry_stream.rs
@@ -3,13 +3,15 @@
 //! real-time access to entries.
 
 use crate::entry::Entry;
-use crate::leader_scheduler::DEFAULT_TICKS_PER_SLOT;
+use crate::leader_scheduler::LeaderScheduler;
 use crate::result::Result;
 use chrono::Utc;
 use std::io::prelude::*;
 use std::net::Shutdown;
 use std::os::unix::net::UnixStream;
 use std::path::Path;
+use std::sync::Arc;
+use std::sync::RwLock;
 
 pub trait EntryStreamHandler {
     fn stream_entries(&mut self, entries: &[Entry]) -> Result<()>;
@@ -17,11 +19,15 @@ pub trait EntryStreamHandler {
 
 pub struct EntryStream {
     pub socket: String,
+    leader_scheduler: Arc<RwLock<LeaderScheduler>>,
 }
 
 impl EntryStream {
-    pub fn new(socket: String) -> Self {
-        EntryStream { socket }
+    pub fn new(socket: String, leader_scheduler: Arc<RwLock<LeaderScheduler>>) -> Self {
+        EntryStream {
+            socket,
+            leader_scheduler,
+        }
     }
 
     fn socket_write(&mut self, payload: &str) -> Result<()> {
@@ -35,6 +41,7 @@ impl EntryStream {
 
 impl EntryStreamHandler for EntryStream {
     fn stream_entries(&mut self, entries: &[Entry]) -> Result<()> {
+        let ticks_per_slot = self.leader_scheduler.read().unwrap().ticks_per_slot;
         for entry in entries {
             let json = serde_json::to_string(&entry)?;
             let payload = format!(
@@ -48,15 +55,15 @@ impl EntryStreamHandler for EntryStream {
             // on block boundary, emit "block" event which signals that a block is closed
             // TODO: look at this in the context of forks, where the tuple (tick_height, entry.id)
             // might not necessarily map 1:1 to block_id
-            if ((entry.tick_height + 1) % DEFAULT_TICKS_PER_SLOT) == 0 {
+            if ((entry.tick_height + 1) % ticks_per_slot) == 0 {
                 let json = serde_json::to_string(&entry.id)?;
                 let payload = format!(
                     r#"{{"dt":"{}","block":{{"id":{},"tick_height":{},"count":{},"block_height":{}}}}}{}"#,
                     Utc::now().to_rfc3339(),
                     json,
-                    entry.tick_height + 1 - DEFAULT_TICKS_PER_SLOT,
-                    DEFAULT_TICKS_PER_SLOT,
-                    (entry.tick_height + 1) / DEFAULT_TICKS_PER_SLOT,
+                    entry.tick_height + 1 - ticks_per_slot,
+                    ticks_per_slot,
+                    (entry.tick_height + 1) / ticks_per_slot,
                     "\n",
                 );
                 self.socket_write(&payload)?;
@@ -73,7 +80,7 @@ pub struct MockEntryStream {
 
 impl MockEntryStream {
     #[allow(clippy::needless_pass_by_value)]
-    pub fn new(_socket: String) -> Self {
+    pub fn new(_socket: String, _leader_scheduler: Arc<RwLock<LeaderScheduler>>) -> Self {
         MockEntryStream { socket: Vec::new() }
     }
 }
diff --git a/src/replay_stage.rs b/src/replay_stage.rs
index 435b1373..8fbd2460 100644
--- a/src/replay_stage.rs
+++ b/src/replay_stage.rs
@@ -188,7 +188,9 @@ impl ReplayStage {
         ledger_signal_receiver: Receiver<bool>,
     ) -> (Self, EntryReceiver) {
         let (ledger_entry_sender, ledger_entry_receiver) = channel();
-        let mut entry_stream = entry_stream.cloned().map(EntryStream::new);
+        let mut entry_stream = entry_stream
+            .cloned()
+            .map(|socket| EntryStream::new(socket, bank.leader_scheduler.clone()));
 
         let exit_ = exit.clone();
         let t_replay = Builder::new()
-- 
2.17.2 (Apple Git-113)
```

**Comment by sunnygleason** [comment]
@tyera @mvines thank you for the patches, I will try those as I track down
an issue with the slot id not incrementing.

I think I'm doing something similar here, but the leader scheduler doesn't
seem to be incrementing the current slot. Is that expected?

(I'm wondering why the suggested patches place slot computation
arithmetic in this class instead of using the slot calculations from
the scheduler. Is it because of a known issue, or because the bank
or some other class is more authoritative?)

In any case, I've tried to improve the tests which should help us track
the issue down. These are my first rust unit tests so feedback is always
welcome.

File: src/entry_stream.rs [hunk 1]
```diff
@@ -3,58 +3,421 @@
...
+        leader_scheduler: &LeaderScheduler,
+        entries: &[Entry],
+    ) -> Result<()> {
+        for entry in entries {
+            let (_prev_leader_id, prev_slot) = leader_scheduler
+                .get_scheduled_leader(self.last_tick_height)
+                .unwrap();
+            let (curr_leader_id, curr_slot) = leader_scheduler
+                .get_scheduled_leader(entry.tick_height)
+                .unwrap();
```
Comment (sunnygleason) [inline src/entry_stream.rs]: @CriesofCarrots @mvines this is where I'm wondering why the slots are not returned properly; any thoughts?

File: src/entry_stream.rs [hunk 2]
```diff
@@ -3,58 +3,421 @@
...
+    use serde_json::Value;
+    use solana_sdk::hash::Hash;
+
+    #[test]
+    fn test_entry_stream() -> () {
+        // Set up entry stream
+        let mut entry_stream = MockEntryStream::new("test_stream".to_string());
+        let bank = Bank::default();
+        let leader_scheduler = bank.leader_scheduler.read().unwrap();
+        let num_entries = 9;
```
Comment (CriesofCarrots) [inline src/entry_stream.rs]: This hard-coded number will make this test fail on the tip of master:
https://github.com/solana-labs/solana/blob/56734dca3b8ba0e5f4973d4767ec3299afdef0be/src/leader_scheduler.rs#L24

File: src/entry_stream.rs [hunk 3]
```diff
@@ -3,58 +3,421 @@
...
+    use crate::leader_scheduler::DEFAULT_TICKS_PER_SLOT;
+    use chrono::{DateTime, FixedOffset};
+    use serde_json::Value;
+    use solana_sdk::hash::Hash;
+
+    #[test]
+    fn test_entry_stream() -> () {
+        // Set up entry stream
+        let mut entry_stream = MockEntryStream::new("test_stream".to_string());
+        let bank = Bank::default();
```
Comment (mvines) [inline src/entry_stream.rs]: The default Bank is sorta useless here.  Try creating a better Bank like this, which will have 5 ticks per slot, and 2 slots per epoch:
```rust
        let leader_scheduler_config = LeaderSchedulerConfig::new(5, 2,10);
        let (genesis_block, mint_keypair) = GenesisBlock::new(1_000_000);
        info!("bootstrap_leader_id: {}", genesis_block.bootstrap_leader_id);
        let bank = Bank::new_with_leader_scheduler_config(&genesis_block, &leader_scheduler_config);
```

File: src/entry_stream.rs (hunk 1 above)
Comment (CriesofCarrots) [reply src/entry_stream.rs]: This method actually doesn't exist anymore on the tip of master, so I'd hate to see you spend anymore of your time troubleshooting it!
I'm working on incorporating @mvines' patches and rebasing your changes. Should have something helpful for you soon.

File: src/entry_stream.rs (hunk 3 above)
Comment (sunnygleason) [reply src/entry_stream.rs]: Excellent, will try that!

File: src/entry_stream.rs (hunk 2 above)
Comment (sunnygleason) [reply src/entry_stream.rs]: Ok, will update!

File: src/entry_stream.rs (hunk 1 above)
Comment (CriesofCarrots) [reply src/entry_stream.rs]: @sunnygleason , I have some code that works with real nodes on the tip of master. Working on updating your test now.

**Pull request closed without merging**